lib/rust/ensogl/core/src/debug/stats.rs

```diff
@@ -15,48 +15,137 @@
+    rc: Rc<RefCell<StatsCollector<T>>>,
 }
 
-impl Default for Stats {
-    fn default() -> Self {
-        let rc = Rc::new(RefCell::new(default()));
-        Self { rc }
+impl<T: TimeProvider> Clone for StatsOverTime<T> {
+    fn clone(&self) -> Self {
+        Self { rc: self.rc.clone() }
     }
 }
 
-impl Stats {
+impl<T: TimeProvider> StatsOverTime<T> {
+    /// Constructor.
+    pub fn new(time_provider: T) -> Self {
+        let stats_collector = StatsCollector::new(time_provider);
+        let rc = Rc::new(RefCell::new(stats_collector));
+        Self { rc }
+    }
+
     /// Starts tracking data for a new animation frame.
-    /// Also, calculates the `fps` stat and updates `frame_counter`.
+    /// Also, calculates the [`fps`] stat.
     /// Returns a snapshot of statistics data for the previous frame.
-    /// Note: on first ever frame, there was no "previous frame", so all returned stats are zero
-    /// (this special case can be recognized by checking `frame_counter == 0`).
-    pub fn begin_frame(&self, time: f64) -> StatsData {
-        self.rc.borrow_mut().begin_frame(time)
+    ///
+    /// Note: on first ever frame, there was no "previous frame", so no data will be returned.
+    ///
+    /// Note: the code works under an assumption that [`begin_frame()`] and [`end_frame()`] are
+    /// called properly on every frame (behavior in case of missed frames or missed calls is not
+    /// specified).
+    pub fn begin_frame(&self) -> Option<StatsData> {
+        self.rc.borrow_mut().begin_frame()
     }
 
     /// Ends tracking data for the current animation frame.
     /// Also, calculates the `frame_time` and `wasm_memory_usage` stats.
-    pub fn end_frame(&self, time: f64) {
-        self.rc.borrow_mut().end_frame(time);
+    pub fn end_frame(&self) {
+        self.rc.borrow_mut().end_frame();
     }
+}
+
+
+
+// ======================
+// === StatsCollector ===
+// ======================
+
+#[derive(Debug)]
+struct StatsCollector<T: TimeProvider> {
+    time_provider:    T,
+    stats_data:       StatsData,
+    frame_begin_time: Option<f64>,
+}
 
-    /// Resets the per-frame statistics.
-    pub fn reset_per_frame_statistics(&self) {
-        self.rc.borrow_mut().reset_per_frame_statistics();
+impl<T: TimeProvider> StatsCollector<T> {
+    /// Constructor.
+    fn new(time_provider: T) -> Self {
+        let stats_data = default();
+        let frame_begin_time = None;
+        Self { time_provider, stats_data, frame_begin_time }
+    }
+
+    fn begin_frame(&mut self) -> Option<StatsData> {
+        let time = self.time_provider.now();
+        let new_frame = StatsData {
+            draw_call_count: 0,
+            shader_compile_count: 0,
+            data_upload_count: 0,
+            data_upload_size: 0,
+            ..self.stats_data
+        };
+        let mut previous_frame = mem::replace(&mut self.stats_data, new_frame);
+        match self.frame_begin_time.replace(time) {
```

Comment:
this can be implemented as a mapping for Option, which we always prefer instead of writing `None => None`.